x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authc/service/ServiceAccountTokenTests.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.core.security.authc.service;
+
+import org.elasticsearch.test.ESTestCase;
+
+import java.util.Arrays;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static org.hamcrest.Matchers.is;
+
+public class ServiceAccountTokenTests extends ESTestCase {
+
+    private static final Set<Character> VALID_TOKEN_NAME_CHARS = Set.of(
+        '0', '1', '2', '3', '4', '5', '6', '7', '8', '9',
+        'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O',
+        'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z',
+        'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o',
+        'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z',
+        '-', '_'
+    );
+
+    private static final Set<Character> INVALID_TOKEN_NAME_CHARS = Set.of(
+        '!', '"', '#', '$', '%', '&', '\'', '(', ')', '*', '+', ',', '.', '/', ':', ';', '<', '=', '>', '?', '@', '[',
+        '\\', ']', '^', '`', '{', '|', '}', '~', ' ', '\t', '\n', '\r');
+
+    public void testIsValidTokenName() {
+        final String tokenName1 = randomTokenName();
+        assertThat(ServiceAccountToken.isValidTokenName(tokenName1), is(true));
+
+        final String tokenName2 = "_" + randomTokenName().substring(1);
+        assertThat(ServiceAccountToken.isValidTokenName(tokenName2), is(false));
+
+        assertThat(ServiceAccountToken.isValidTokenName(null), is(false));
+
+        final String tokenName3 = randomInvalidTokenName();
+        assertThat(ServiceAccountToken.isValidTokenName(tokenName3), is(false));
+    }
+
```

Comment:
Do we not also need tests for building/parsing the bearer token?

Reply:
I have those in `ServiceAccountServiceTests`. But I agree we could use more tests. So I added them to both here and `ServiceAccountServiceTests`